### ISSUE TITLE
Unstable training with larger batch size or multi-GPU

### ISSUE THREAD

**Issue by Dustbin-Li** [zhangxjohn/MoDr#1]
Hi, thank you for your insightful work!

I was trying to reproduce your results, but I encountered an issue when increasing the batch size. Specifically:
- When changing the batch size from 8 to 24 on a single GPU, or using batch size 16 on 4 GPUs, the training becomes extremely unstable.
- The loss increases significantly, and the gradient norm (grad_norm) sometimes becomes NaN.

Here are some screenshots illustrating the issue:

<img width="658" height="380" alt="Image" src="https://github.com/user-attachments/assets/99829e2f-5b72-4747-851f-2fe65e118d6f" />

<img width="658" height="380" alt="Image" src="https://github.com/user-attachments/assets/10a56ece-5c87-4a30-8065-e931d18a0310" />

![Image](https://github.com/user-attachments/assets/83bbea7d-8d39-4867-a4d1-0f5ab433689a)

Could you provide any guidance on why this might be happening and how to stabilize training for larger batch sizes or multi-GPU setups?

**Comment by zhangxjohn** [comment]
Thanks for trying out our work!

Regarding your issue, due to Huginn’s recurrent structure, the released checkpoint doesn’t support multi-GPU training, so all our experiments were run strictly on a single GPU. The NaN gradients you saw on 4 GPUs are likely caused by this limitation.

Also, on a single GPU, we only used small batch sizes (4 or 8). When we tried increasing the batch size ourselves, the training would usually just crash directly due to OOM (Out-Of-Memory) rather than showing loss spikes.

These are our observations from our experiments. We provide this information just for your reference.
